pkg/io/video/scaler_cgo.go

```diff
@@ -92,3 +110,74 @@ func (f *FastNearestNeighbor) Scale(dst draw.Image, dr image.Rectangle, src imag
 		panic("unimplemented")
 	}
 }
+
+// FastBoxSampling is a CGO version of Box sampling scaler.
+type FastBoxSampling struct {
+}
+
+// Scale implements the draw.Scaler interface.
+func (f *FastBoxSampling) Scale(dst draw.Image, dr image.Rectangle, src image.Image, sr image.Rectangle, op draw.Op, opts *draw.Options) {
+	ddx, ddy := dr.Dx(), dr.Dy()
+	sdx, sdy := sr.Dx(), sr.Dy()
+	if sdx < ddx && sdy < ddy {
+		// Box upsampling is equivalent of NearestNeighbor
+		(&FastNearestNeighbor{}).Scale(dst, dr, src, sr, op, opts)
+		return
+	}
+	tmp := poolBoxSampleBuffer.Get().(*[]uint32)
+	defer poolBoxSampleBuffer.Put(tmp)
+
+	switch s := src.(type) {
+	case (*image.RGBA):
+		d := dst.(*image.RGBA)
+		l := d.Stride * d.Rect.Dy()
+		if len(d.Pix) < l {
+			if cap(d.Pix) < l {
+				d.Pix = make([]uint8, l)
+			}
+			d.Pix = d.Pix[:l]
+		}
+		if cap(*tmp) < l {
+			*tmp = make([]uint32, l)
+		}
+		C.fastBoxSampling(
+			(*C.uchar)(&d.Pix[dr.Min.X+d.Stride*dr.Min.Y]),
+			(*C.uchar)(&s.Pix[sr.Min.X+s.Stride*sr.Min.Y]),
+			4,
+			C.int(dr.Dx()), C.int(dr.Dy()), C.int(d.Stride),
+			C.int(sr.Dx()), C.int(sr.Dy()), C.int(s.Stride),
+			(*C.uint32_t)(&(*tmp)[0]),
+		)
+	case (*image.Gray):
+		d := dst.(*image.Gray)
+		l := d.Stride * d.Rect.Dy()
+		if len(d.Pix) < l {
+			if cap(d.Pix) < l {
+				d.Pix = make([]uint8, l)
+			}
+			d.Pix = d.Pix[:l]
+		}
+		if cap(*tmp) < l {
+			*tmp = make([]uint32, l)
+		}
```

Comment:
I think `len(*tmp)` makes more sense here. I feel like it's a little bit weird if let say, we have a len of 10 and a cap of 20. Since C has access to the whole memory, 0 to 19, C can technically just write to it. And, from Go, when we append the slice, the new memory would be non-zero.

Reply:
I had intended to avoid reallocation when length is smaller but capacity is enough. However, this temporary buffer is only used by this function and length is never trimmed, so `len()` would be fine.